### ISSUE TITLE
Failed deploy reported successful

### ISSUE THREAD

**Issue by stefanmb** [Shopify/kubernetes-deploy#20]
Syntax errors in the k8s template error out, but the deploy still reports as successful:

```
[WARN]	The following command failed: kubectl apply -f /tmp/testing-deployment.yml.erb20170209-12190-s9f69e --prune --all --prune-whitelist\=core/v1/ConfigMap --prune-whitelist\=core/v1/Pod --prune-whitelist\=core/v1/Service --prune-whitelist\=batch/v1/Job --prune-whitelist\=extensions/v1beta1/DaemonSet --prune-whitelist\=extensions/v1beta1/Deployment --prune-whitelist\=extensions/v1beta1/HorizontalPodAutoscaler --prune-whitelist\=extensions/v1beta1/Ingress --prune-whitelist\=apps/v1beta1/StatefulSet --namespace\=pipa-test --context\=pipa-test
[WARN]	error: unable to decode "/tmp/testing-deployment.yml.erb20170209-12190-s9f69e": [pos 290]: json: expect char '"' but got char '8'
Waiting for Deployment/pipa-test
[KUBESTATUS] {"group":"Pipa-test deployment","name":"pipa-test-1149558948-5l604","status_string":"Running (Ready: true)","exists":true,"succeeded":true,"failed":false,"timed_out":false}
...
Spent 0.58s waiting for Deployment/pipa-test
Deploy succeeded!
```

Should this error be caught in `shipit-engine` or in this gem?

**Comment by KnVerey** [comment]
It's the gem's problem. Thanks for reporting.

**Comment by kirs** [comment]
@KnVerey I'm wondering if we should raise the error here: https://github.com/Shopify/kubernetes-deploy/blob/1da3249/lib/kubernetes-deploy/runner.rb#L305

**Comment by KnVerey** [comment]
No, because non-fatal errors can be seen there too. That method passes back the status, and the caller is supposed to deal with it. The weird thing here is that the initial dry run did not raise: https://github.com/Shopify/kubernetes-deploy/blob/1da3249/lib/kubernetes-deploy/runner.rb#L145. That implies it was the load->dump that introduced the yaml error, which is pretty strange.

In any case, the thing to do IMO is to handle the result here: https://github.com/Shopify/kubernetes-deploy/blob/1da3249/lib/kubernetes-deploy/runner.rb#L271. I.e. on success, log the output so they can see what was pruned, and raise on failure.

**Comment by kirs** [comment]
Reproduced it with by YAML that would be invalid from the k8s point of view (but valid for Ruby parser):

```diff
--- a/test/fixtures/basic/web.yml.erb
+++ b/test/fixtures/basic/web.yml.erb
@@ -3,7 +3,10 @@ kind: Ingress
 metadata:
   name: web
   labels:
-    name: web
+    name:
+      not_a_name:
+        - 1
+        - 2
```

**Comment by kirs** [comment]
And then `kubectl create --dry-run` considers this file valid.
Later, `kubectl create` without `--dry-run` would fail with `[pos 96]: json: expect char '"' but got char '{'`.

Let's track the issue here: https://github.com/kubernetes/kubernetes/issues/42057

**Comment by kirs** [comment]
Fixed in #31
